utils/utils.go

```diff
@@ -0,0 +1,62 @@
+package utils
+
+import (
+	"net"
+	"strconv"
+
+	"github.com/gravitational/trace"
+)
+
+// PortList is a list of TCP ports
+type PortList []string
+
+// Pop returns a value from the list, it panics if the value is not there
+func (p *PortList) Pop() string {
+	if len(*p) == 0 {
+		panic("list is empty")
+	}
+	val := (*p)[len(*p)-1]
+	*p = (*p)[:len(*p)-1]
+	return val
+}
+
+// PopInt returns a value from the list, it panics if not enough values
+// were allocated
+func (p *PortList) PopInt() int {
+	i, err := strconv.Atoi(p.Pop())
+	if err != nil {
+		panic(err)
+	}
+	return i
+}
+
+// PopIntSlice returns a slice of values from the list, it panics if not enough
+// ports were allocated
+func (p *PortList) PopIntSlice(num int) []int {
+	ports := make([]int, num)
+	for i := range ports {
+		ports[i] = p.PopInt()
+	}
+	return ports
+}
+
+// GetFreeTCPPortsForTests returns n free ports (which are suggested by the kernel)
```

Comment:
Does it really help running tests in parallel? I see there's still a chance for race condition.

Ideally, the issue should be solved if both Teleport auth server and plugin http server could listen to `:0` and somehow report the allocated port to the app. I even have a half-working prototype for this.

Perhaps leave it alone for now and run tests with `go test -parallel 1`?

Reply:
This solution doesn't fix the problem, but makes it much less likely.
Actually listening on port 0 would certainly be preferable.

I added some code in https://github.com/gravitational/teleport/pull/3713/commits/13c0b8aceb59c2d9715e44706120746ba1519d4a to expose actual ports from `TeleportProcess` if it helps.

Reply:
My goal with this PR was basically to fix the unit tests to the point where they would run in CI without error the majority of the time (unless there was a bug introduced by a PR) which is now the case.

As I understand it, Teleport's use of `gocheck` pretty much prevents us from running tests in parallel anyway so the `-parallel` flag wouldn't really make any difference? Certainly, I ran `go test -count 1 ./...` and `go test -count 1 -parallel 1 ./...` a few times and nothing seemed to be different between the two at all.

@marshall-lee If you'd like to fix the problem 'properly' in future by actually passing Teleport's listening port through to the tests that would be awesome. For now, I'd like to get this solution merged so we have 'working' CI.

Reply:
`-parallel` affects the number of tests *within* a package, if they call `t.Parallel()` (not a thing with gocheck).
`-p` affects the number of *different* packages tested concurrently, which is the source of these race conditions.

You can use `go test -p=1` as a bandaid for CI while a proper fix is in progress.
Note that it also prevents parallel compilation, so the whole CI pipeline will be slower with this.

Reply:
@webvictim do you you need a review from anyone else or I can merge this branch?

Reply:
@marshall-lee It's all good now, I'll merge it!